File: backend/app/models/job.py
"""作业域模型：SSH 凭据、可复用流程模板和工单模板。"""

from datetime import datetime

from sqlalchemy import Boolean, Index, Integer, JSON, String, Text, UniqueConstraint
from sqlalchemy.dialects.mysql import MEDIUMTEXT
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, UBIGINT, created_at_column, pk_column, updated_at_column

MTEXT = MEDIUMTEXT().with_variant(Text(), "sqlite")


class Credential(Base):
    """独立管理 SSH 登录凭据；模板只引用作业主机，不直接引用凭据。"""

    __tablename__ = "credential"

    id: Mapped[int] = pk_column()
    name: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    login_user: Mapped[str] = mapped_column(String(64), nullable=False)
    auth_type: Mapped[str] = mapped_column(String(16), nullable=False)
    secret_enc: Mapped[str] = mapped_column(Text, nullable=False)
    passphrase_enc: Mapped[str | None] = mapped_column(Text)
    description: Mapped[str | None] = mapped_column(String(255))
    created_by: Mapped[int | None] = mapped_column(UBIGINT)
    created_at: Mapped[datetime] = created_at_column()
    updated_at: Mapped[datetime] = updated_at_column()


class TicketTemplate(Base):
    """工单业务入口：绑定流程并独立维护本入口的参数契约。"""

    __tablename__ = "ticket_template"
    __table_args__ = (
        Index("idx_ticket_template_jh", "job_host_id"),
        Index("idx_ticket_template_process", "process_template_id"),
        Base.__table_args__,
    )

    id: Mapped[int] = pk_column()
    name: Mapped[str] = mapped_column(String(128), unique=True, nullable=False)
    type: Mapped[str] = mapped_column(String(16), nullable=False)
    description: Mapped[str | None] = mapped_column(String(512))
    job_host_id: Mapped[int] = mapped_column(UBIGINT, nullable=False)
    process_template_id: Mapped[int] = mapped_column(UBIGINT, nullable=False)
    params_schema: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    generator_script: Mapped[str | None] = mapped_column(MTEXT)
    generator_timeout: Mapped[int | None] = mapped_column(Integer)
    allow_withdraw: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    notify_rules: Mapped[list | None] = mapped_column(JSON)
    visible_role_ids: Mapped[list | None] = mapped_column(JSON)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="enabled")
    created_by: Mapped[int | None] = mapped_column(UBIGINT)
    created_at: Mapped[datetime] = created_at_column()
    updated_at: Mapped[datetime] = updated_at_column()


class ProcessTemplate(Base):
    """可复用流程：只维护步骤、执行策略和步骤审批，不绑定参数契约。"""

    __tablename__ = "process_template"
    __table_args__ = (
        Index("idx_process_template_status", "status"),
        Base.__table_args__,
    )

    id: Mapped[int] = pk_column()
    name: Mapped[str] = mapped_column(String(128), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(String(512))
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="enabled")
    exec_strategy: Mapped[dict] = mapped_column(JSON, nullable=False, default=dict)
    created_by: Mapped[int | None] = mapped_column(UBIGINT)
    created_at: Mapped[datetime] = created_at_column()
    updated_at: Mapped[datetime] = updated_at_column()


class ProcessStep(Base):
    """流程步骤：严格串行，每步最多绑定一个前置审批角色。"""

    __tablename__ = "process_step"
    __table_args__ = (
        UniqueConstraint("process_template_id", "step_order", name="uk_process_step"),
        Index("idx_process_step_process", "process_template_id"),
        Base.__table_args__,
    )

    id: Mapped[int] = pk_column()
    process_template_id: Mapped[int] = mapped_column(UBIGINT, nullable=False)
    step_order: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    script_type: Mapped[str] = mapped_column(String(16), nullable=False)
    content: Mapped[str] = mapped_column(MTEXT, nullable=False)
    timeout: Mapped[int] = mapped_column(Integer, nullable=False, default=600)
    approval_role_id: Mapped[int | None] = mapped_column(UBIGINT)

File: frontend/src/api/job.ts
// 凭据 / 工单模板 API（04-API设计 §5，V2）
import { request } from './http'
import type { PageResult } from './system'
import type { ExecStrategy } from './ticket'

// ---------- 凭据（密文任何接口不回显，PRD CRED-01） ----------

export interface CredentialItem {
  id: number
  name: string
  login_user: string | null
  auth_type: 'password' | 'private_key' | 'api_token' | 'username_password' | 'secret_file'
  has_passphrase: boolean
  file_name: string | null
  description: string | null
  created_at: string | null
  updated_at: string | null
}

/** 凭据表单：secret 编辑时留空 = 不变更密文 */
export interface CredentialForm {
  name: string
  login_user?: string | null
  auth_type: CredentialItem['auth_type']
  secret: string
  passphrase?: string
  file_name?: string
  description?: string
}

export function listCredentials(params: {
  page?: number
  page_size?: number
  keyword?: string
  auth_type?: string
}) {
  return request<PageResult<CredentialItem>>({ url: '/credentials', method: 'get', params })
}

export function createCredential(data: CredentialForm) {
  return request<{ id: number }>({ url: '/credentials', method: 'post', data })
}

export function updateCredential(id: number, data: CredentialForm) {
  return request<null>({ url: `/credentials/${id}`, method: 'put', data })
}

export function deleteCredential(id: number) {
  return request<null>({ url: `/credentials/${id}`, method: 'delete' })
}

// ---------- 工单模板（V2：全量规则配置，规则任一变更自动升版 TPL-06） ----------

export type TemplateType = 'release' | 'daily_ops' | 'other'
export type TemplateStatus = 'enabled' | 'disabled'
export type ScriptType = 'shell' | 'playbook'
export type TicketParamSource = 'fixed' | 'user' | 'generated'
export type TicketInputType = 'text' | 'enum'

export interface TicketParam {
  name: string
  label?: string | null
  source: TicketParamSource
  input_type: TicketInputType
  options: string[]
  default?: string | null
  required: boolean
  description?: string | null
}

export interface ProcessStep {
  step_order?: number
  name: string
  script_type: ScriptType
  content: string
  timeout: number
  approval_role_id?: number | null
}

export interface ProcessTemplateItem {
  id: number
  name: string
  description: string | null
  status: TemplateStatus
  exec_strategy: Partial<ExecStrategy>
  steps_count?: number | null
  ticket_template_refs?: number | null
  created_at: string | null
  updated_at: string | null
}

export interface ProcessTemplateForm {
  name: string
  description?: string
  exec_strategy: ExecStrategy
  steps: ProcessStep[]
}

export interface TicketTemplateForm {
  name: string
  type: TemplateType
  description?: string
  job_host_id: number
  process_template_id: number
  params_schema: TicketParam[]
  generator_script?: string | null
  generator_timeout?: number | null
  credential_refs: CredentialRef[]
  allow_withdraw: boolean
  notify_rules: NotifyRule[]
  visible_role_ids: number[]
  status?: TemplateStatus
}

export interface CredentialRef {
  alias: string
  credential_id: number
  credential_name?: string | null
}

export function defaultExecStrategy(): ExecStrategy {
  return { timeout: 600, fail_fast: true, kill_on_stop: false }
}

export function listProcessTemplates(params: { page?: number; page_size?: number; keyword?: string; status?: string } = {}) {
  return request<PageResult<ProcessTemplateItem>>({ url: '/process-templates', method: 'get', params })
}

export function getProcessTemplate(id: number) {
  return request<ProcessTemplateItem & { steps: ProcessStep[] }>({ url: `/process-templates/${id}`, method: 'get' })
}

export function createProcessTemplate(data: ProcessTemplateForm) {
  return request<{ id: number }>({ url: '/process-templates', method: 'post', data })
}

export function updateProcessTemplate(id: number, data: ProcessTemplateForm) {
  return request<{ id: number }>({ url: `/process-templates/${id}`, method: 'put', data })
}

export function setProcessTemplateStatus(id: number, status: TemplateStatus) {
  return request<{ status: TemplateStatus }>({ url: `/process-templates/${id}/status`, method: 'put', data: { status } })
}

export function deleteProcessTemplate(id: number) {
  return request<null>({ url: `/process-templates/${id}`, method: 'delete' })
}

/** 通知规则行：receivers 支持 creator / approver_role / role:<id>（TPL-05） */
export interface NotifyRule {
  event: string
  receivers: string[]
  channels: string[]
}

/** 工单模板列表行：包含业务入口和参数契约。 */
export interface TemplateItem {
  id: number
  name: string
  type: TemplateType
  description: string | null
  job_host_id: number
  process_template_id: number
  params_schema: TicketParam[]
  generator_script?: string | null
  generator_timeout?: number | null
  credential_refs?: CredentialRef[]
  allow_withdraw: boolean
  notify_rules: NotifyRule[]
  visible_role_ids: number[]
  status: TemplateStatus
  created_at: string | null
  updated_at: string | null
}

/** 工单模板详情：业务入口配置和流程引用。 */
export interface TemplateDetail extends TemplateItem {
  process_template?: { id: number; name: string; status: TemplateStatus }
}

export function listTemplates(params: {
  page?: number
  page_size?: number
  keyword?: string
  type?: string
  status?: string
  process_template_id?: number
}) {
  return request<PageResult<TemplateItem>>({ url: '/templates', method: 'get', params })
}

export function getTemplate(id: number) {
  return request<TemplateDetail>({ url: `/templates/${id}`, method: 'get' })
}

export function createTemplate(data: TicketTemplateForm) {
  return request<{ id: number }>({ url: '/templates', method: 'post', data })
}

export function updateTemplate(id: number, data: TicketTemplateForm) {
  return request<{ id: number }>({
    url: `/templates/${id}`,
    method: 'put',
    data,
  })
}

/** 启用/禁用：禁用后不可被提交工单，不影响已提交工单（TPL-03） */
export function setTemplateStatus(id: number, status: TemplateStatus) {
  return request<{ status: TemplateStatus }>({
    url: `/templates/${id}/status`,
    method: 'put',
    data: { status },
  })
}

export function deleteTemplate(id: number) {
  return request<null>({ url: `/templates/${id}`, method: 'delete' })
}
